### PR TITLE
#12538 force RST packets to be re-sent by sending data after the peer aborts in abortConnection tests

### LINKED ISSUE #12538: `twisted.internet.test.test_tcp.AbortConnectionTests_*.test_resumeProducingAbortLater` is still flaky on macOS
Running on my local machine (26.1 / 25B78):

```
twisted.internet.test.test_tcp.AbortConnectionTests_AsyncioSelectorReactorTests.test_resumeProducingAbortLater ... Traceback (most recent call last):
  File "/Users/glyph/Projects/Twisted/src/twisted/internet/test/test_tcp.py", line 2905, in test_resumeProducingAbortLater
    return self.runAbortTest(
  File "/Users/glyph/Projects/Twisted/src/twisted/internet/test/test_tcp.py", line 2816, in runAbortTest
    reactor = runProtocolsWithReactor(self, server, client, self.endpoints)
  File "/Users/glyph/Projects/Twisted/src/twisted/internet/test/connectionmixins.py", line 177, in runProtocolsWithReactor
    reactorBuilder.runReactor(reactor)
  File "/Users/glyph/Projects/Twisted/src/twisted/internet/test/reactormixins.py", line 352, in runReactor
    raise TestTimeoutError(f"reactor still running after {timeout} seconds")
twisted.internet.test.reactormixins.TestTimeoutError: reactor still running after 120.0 seconds
[ERROR]
twisted.internet.test.test_tcp.AbortConnectionTests_AsyncioSelectorReactorTests.test_resumeProducingThrows ... [OK]
twisted.internet.test.test_tcp.AbortConnectionTests_AsyncioSelectorReactorTests.test_serverAbortsConnectionThenLosesConnection ... [OK]
twisted.internet.test.test_tcp.AbortConnectionTests_AsyncioSelectorReactorTests.test_serverAbortsConnectionTwice ... [OK]
twisted.internet.test.test_tcp.AbortConnectionTests_CFReactorTests.test_clientAbortsConnectionThenLosesConnection ... [OK]
twisted.internet.test.test_tcp.AbortConnectionTests_CFReactorTests.test_clientAbortsConnectionTwice ... [OK]
twisted.internet.test.test_tcp.AbortConnectionTests_CFReactorTests.test_dataReceivedAbort ... [OK]
twisted.internet.test.test_tcp.AbortConnectionTests_CFReactorTests.test_dataReceivedThrows ... [OK]
twisted.internet.test.test_tcp.AbortConnectionTests_CFReactorTests.test_fullWriteBuffer ... [OK]
twisted.internet.test.test_tcp.AbortConnectionTests_CFReactorTests.test_fullWriteBufferAfterByteExchange ... [OK]
twisted.internet.test.test_tcp.AbortConnectionTests_CFReactorTests.test_resumeProducingAbort ... Traceback (most recent call last):
  File "/Users/glyph/Projects/Twisted/src/twisted/internet/test/test_tcp.py", line 2892, in test_resumeProducingAbort
    self.runAbortTest(ProducerAbortingClient, ConnectableProtocol)
  File "/Users/glyph/Projects/Twisted/src/twisted/internet/test/test_tcp.py", line 2816, in runAbortTest
    reactor = runProtocolsWithReactor(self, server, client, self.endpoints)
  File "/Users/glyph/Projects/Twisted/src/twisted/internet/test/connectionmixins.py", line 177, in runProtocolsWithReactor
    reactorBuilder.runReactor(reactor)
  File "/Users/glyph/Projects/Twisted/src/twisted/internet/test/reactormixins.py", line 352, in runReactor
    raise TestTimeoutError(f"reactor still running after {timeout} seconds")
twisted.internet.test.reactormixins.TestTimeoutError: reactor still running after 120.0 seconds
[ERROR]
```

I haven't seen these tests fail for a while, and have in fact been successfully running full test suites for a while; this may be new behavior in Tahoe or indeed specifically in 26.1.  But, I see two `skipIf`s on these tests which say they are "Flaky on macOS on Azure" so I am not sure they're being covered in CI at all right now.

### PR DESCRIPTION
**Pull request by glyph**
## Scope and purpose

Fixes #12538

### CODE REVIEW

**Comment by glyph** [comment]
@graingert OK so I had a bit of an existential crisis thinking about what exactly the *true meaning* of an RST TCP packet really was, and I realized this:

While it is deeply annoying behavior for macOS to fail to process RST packets like this, it's also not exactly *wrong*.  If these integration tests were operating over anything besides a loopback interface, and some router in the middle dropped the individual `SO_LINGER`-zero RST packet, we would consider it within its rights to do so. Thus, the fact that an RST packet got sent but not processed over lo0 is … also not totally wrong.  When you set `SO_LINGER` to zero, you are specifically saying "discard the resources that would allow you to promptly tear down the connection in the face of packet loss, i need them back _right now_".

So, if you're running a server and you shut down all your connections, and it delivers all their RSTs in a rush, that means that the clients might be notified in a timely way, and that's exactly the behavior you've requested.

Thus, rather than doing something macOS specific or skipping tests here, this introduces a more comprehensive approach of adding explicit `transport.write()` on the potentially-idling peer after every `abortConnection` in the relevant tests.

With this change, I'm easily able to locally get 50 passes via `trial -e -u twisted.internet.test.test_tcp`, whereas previously I'd be lucky to get through 3.

File: src/twisted/protocols/tls.py [hunk 1]
```diff
@@ -642,7 +643,7 @@ def write(self, data: bytes) -> None:
 
         Accumulating too much data can result in higher memory usage.
         """
-        self._buffer.append(data)
+        self._buffer.append(data + b"")
```
Comment (gudnimg) [inline src/twisted/protocols/tls.py:646]: Is this change really needed? Its not obvious to me what this extra `b""` does. Does it slow down the code?

`data` is typed as `bytes` 🤔

Comment (glyph) [reply src/twisted/protocols/tls.py:646]: There's a test that verifies that we get a `TypeError` when writing non-`bytes` to the transport. The test was kind of janky and verifying a logged message rather than actually getting an exception, so this is a bit of a cleanup that makes it immediate. Doing it as `+ b""` is faster than doing an `isinstance` check in Python.

It doesn't have a material impact on any of the benchmarks.

Comment (glyph) [reply src/twisted/protocols/tls.py:646]: Oh, and the reason that I needed to touch this in the first place is that a small perturbation in the async scheduling and writing data in the tcp tests resulted in the TLS timer failing to get cleaned up; and this was a general reliability issue with this transport, there's no `Deferred` you can hang on to in order to ensure that the TLS transport callback is finished

**Comment by glyph** [comment]
I left the tests running overnight and got up through:

```
PASSED (skips=87, successes=384)
Test Pass 232
```

after running for 29678 seconds.

on `trial -e -u twisted.internet.test.test_tcp`.  It did eventually fail on:

```
===============================================================================
[FAIL]
Traceback (most recent call last):
  File "/Users/glyph/Projects/Twisted/src/twisted/internet/test/test_tcp.py", line 1396, in test_closePeerOnEMFILE
    assertPeerClosedOnEMFILE(
  File "/Users/glyph/Projects/Twisted/src/twisted/internet/test/test_tcp.py", line 1253, in assertPeerClosedOnEMFILE
    testCase.assertNoResult(clientFactory.failDeferred)
  File "/Users/glyph/Projects/Twisted/src/twisted/trial/_synctest.py", line 821, in assertNoResult
    self.fail(
twisted.trial.unittest.FailTest: No result expected on <Deferred at 0x73e20a3cd0 current result: None>, found None instead

twisted.internet.test.test_tcp.TCPFDPortTestsBuilder_AsyncioSelectorReactorTests.test_closePeerOnEMFILE
===============================================================================
[FAIL]
Traceback (most recent call last):
  File "/Users/glyph/Projects/Twisted/src/twisted/internet/test/test_tcp.py", line 1396, in test_closePeerOnEMFILE
    assertPeerClosedOnEMFILE(
  File "/Users/glyph/Projects/Twisted/src/twisted/internet/test/test_tcp.py", line 1253, in assertPeerClosedOnEMFILE
    testCase.assertNoResult(clientFactory.failDeferred)
  File "/Users/glyph/Projects/Twisted/src/twisted/trial/_synctest.py", line 821, in assertNoResult
    self.fail(
twisted.trial.unittest.FailTest: No result expected on <Deferred at 0x73e2b5bdd0 current result: None>, found None instead

twisted.internet.test.test_tcp.TCPFDPortTestsBuilder_CFReactorTests.test_closePeerOnEMFILE
===============================================================================
[FAIL]
Traceback (most recent call last):
  File "/Users/glyph/Projects/Twisted/src/twisted/internet/test/test_tcp.py", line 1396, in test_closePeerOnEMFILE
    assertPeerClosedOnEMFILE(
  File "/Users/glyph/Projects/Twisted/src/twisted/internet/test/test_tcp.py", line 1253, in assertPeerClosedOnEMFILE
    testCase.assertNoResult(clientFactory.failDeferred)
  File "/Users/glyph/Projects/Twisted/src/twisted/trial/_synctest.py", line 821, in assertNoResult
    self.fail(
twisted.trial.unittest.FailTest: No result expected on <Deferred at 0xf3fbbf950 current result: None>, found None instead

twisted.internet.test.test_tcp.TCPFDPortTestsBuilder_SelectReactorTests.test_closePeerOnEMFILE
===============================================================================
[FAIL]
Traceback (most recent call last):
  File "/Users/glyph/Projects/Twisted/src/twisted/internet/test/test_tcp.py", line 1396, in test_closePeerOnEMFILE
    assertPeerClosedOnEMFILE(
  File "/Users/glyph/Projects/Twisted/src/twisted/internet/test/test_tcp.py", line 1253, in assertPeerClosedOnEMFILE
    testCase.assertNoResult(clientFactory.failDeferred)
  File "/Users/glyph/Projects/Twisted/src/twisted/trial/_synctest.py", line 821, in assertNoResult
    self.fail(
twisted.trial.unittest.FailTest: No result expected on <Deferred at 0x73e5953dd0 current result: None>, found None instead

twisted.internet.test.test_tcp.TCPFDPortTestsBuilder_ThreadedSelectReactorTests.test_closePeerOnEMFILE
===============================================================================
[FAIL]
Traceback (most recent call last):
  File "/Users/glyph/Projects/Twisted/src/twisted/internet/test/test_tcp.py", line 1396, in test_closePeerOnEMFILE
    assertPeerClosedOnEMFILE(
  File "/Users/glyph/Projects/Twisted/src/twisted/internet/test/test_tcp.py", line 1253, in assertPeerClosedOnEMFILE
    testCase.assertNoResult(clientFactory.failDeferred)
  File "/Users/glyph/Projects/Twisted/src/twisted/trial/_synctest.py", line 821, in assertNoResult
    self.fail(
twisted.trial.unittest.FailTest: No result expected on <Deferred at 0x7466cbcf50 current result: None>, found None instead

twisted.internet.test.test_tcp.TCPPortTestsBuilder_AsyncioSelectorReactorTests.test_closePeerOnEMFILE
-------------------------------------------------------------------------------
Ran 471 tests in 698.173s

FAILED (skips=87, failures=5, successes=379)
```

but that might have just been an unavoidable resource exhaustion when running the tests for that long.

Previously I could only get the tests to pass three or four times in a row.

**Comment by glyph** [comment]
OK, I think that while this sadly has not fixed *all* of our flaky tests and build reliability problems, it *is* a massive improvement.

**Comment by glyph** [comment]
please review

File: src/twisted/protocols/tls.py (hunk 1 above)
Comment (adiroiban) [reply src/twisted/protocols/tls.py:646]: Maybe have an inline comment to explain why this is needed.

As a general note. When a reviewer asked if a code is needed, maybe the answer should be inside the code (brief comment or link to PR comment), not only as PR reply. In this way, the information is preserved outside of the PR process.

File: src/twisted/internet/test/connectionmixins.py [hunk 2]
```diff
@@ -71,7 +76,25 @@ class ConnectableProtocol(Protocol):
 
     disconnectReason = None
 
-    def _setAttributes(self, reactor, done):
+    def goodbye(self) -> None:
```
Comment (adiroiban) [inline src/twisted/internet/test/connectionmixins.py]: Minor comment

Maybe call it . `forceRSTDelivery` or `forceRST`

I feel that goodbye is just a cute name, that doesn't convey the purpose of this code.

**Review by adiroiban — APPROVED** [review]
Great work here!

Happy to see the macOS tests re-enabled.

It's good to run tests on macOS since together with Linux and Windows they can help detect all kind of corner case in terms of TCP behaviour.

File: src/twisted/protocols/tls.py (hunk 1 above)
Comment (gudnimg) [reply src/twisted/protocols/tls.py:646]: > Doing it as + b"" is faster than doing an isinstance check in Python.

This would be great to have as a comment :) It gives at least some context why `b""` is used at all.

File: src/twisted/internet/test/connectionmixins.py (hunk 2 above)
Comment (gudnimg) [reply src/twisted/internet/test/connectionmixins.py]: I agree, the name stood out to me, but I didn't have a better suggestion. I like your suggestion @adiroiban

Comment (glyph) [reply src/twisted/internet/test/connectionmixins.py]: I don't love the name because the actual behavior being triggered does not need to be a *literal* RST packet, particularly if this ever gets used with a different type of transport; it may be other types of connection state negotiation happening internally; we don't really have any way of knowing. However, you're right that the name was overly cute and I've updated it to be clearer.

**Comment by glyph** [comment]
it will be very funny if this commit causes a CI failure

**Review by gudnimg — APPROVED** [review]
